curated_transformers/models/embeddings.py

```diff
@@ -29,3 +29,84 @@ def forward(self, x: Tensor) -> Tensor:
+
+
+class RotaryEmbeddings(Module):
+    cos: Tensor
+    sin: Tensor
+    theta: Tensor
+
+    def __init__(self, width: int, *, seq_len: int = 512, base: int = 10000):
+        """Rotary embeddings (Su et al., 2021) layer. The rotary embedding
+        will be precomputed for up to 'seq _len' positions. The embedding
+        will be recomputed when a longer sequence is found in the input.
+
+        width (int):
+            Rotary embedding dimensionality, must be even.
+        seq_len (int):
+            Number of positons to initially precompute.
+        base (int):
+            The base used for Θ_i, determines the cycle length of the
+            embeddings."""
+        super().__init__()
+
+        if width % 2:
+            raise ValueError(f"Width of rotary embeddings must be even, was: {width}")
+
+        # Θ_i = 10000^(-2(i-1)/d)
+        theta = torch.pow(base, -torch.arange(0, width, 2, dtype=torch.float) / width)
+        self.register_buffer("theta", theta, persistent=False)
+
+        self._create_rotary_embed(width=width, length=seq_len)
+
+    def _create_rotary_embed(self, *, width: int, length: int):
+        # mΘ
+        position = torch.arange(length).unsqueeze(1)
+        m_theta = position * self.theta.unsqueeze(0)
+
+        # We apply both sin and cos twice (see Eq 15, 34), but the ordering
+        # is changed for compatibility with most common implementations.
+        m_theta = torch.cat([m_theta, m_theta], dim=-1)
+
+        re_cos = m_theta.cos().view([length, width])
+        re_sin = m_theta.sin().view([length, width])
+
+        self.register_buffer("cos", re_cos, persistent=False)
+        self.register_buffer("sin", re_sin, persistent=False)
+
+    def _rotate(self, input: Tensor):
+        """Rotate the input tensor by half of its innermost width.
+
+        input (Tensor): array to rotate.
+        RETURNS (Tensor): rotated array.
+
+        Shapes:
+            input - (..., width)
+            output - (..., width)
+        """
+        half_idx = input.shape[-1] // 2
+        input_1 = -input[..., half_idx:]
+        input_2 = input[..., :half_idx]
+        return torch.cat([input_1, input_2], dim=-1)
+
+    def forward(self, input: torch.Tensor):
+        """
+        Apply rotary embeddings to an array.
+
+        input (Tensor): Array to apply the rotary embeddings to.
+        RETURNS (Tensor]: Array with the rotary embeddings applied.
+
+        Shapes:
+            input - (..., seq_len, width_per_head)
+            output - pair of (..., seq_len, width_per_head)
+        """
+        seq_len = input.size(-2)
+        width = self.cos.size(-1)
+        if self.cos.size(-2) < seq_len:
+            self._create_rotary_embed(width=width, length=seq_len)
+        shape = ((1,) * (input.ndim - self.cos.ndim)) + (seq_len, width)
+        rot_cos = self.cos[:seq_len, :].view(shape)
+        rot_sin = self.sin[:seq_len, :].view(shape)
+
+        # Eq 34 with ordering changed for compatibility.
```

Comment:
Do we know the reason behind the change in ordering?

Reply:
I think this precedes the `transformers` implementation, it is also done this way in the original GPT-NeoX codebase:

https://github.com/EleutherAI/gpt-neox/blob/b608043be541602170bfcfb8ec9bf85e8a0799e0/megatron/model/positional_embeddings.py#L38

But I don't think it's really motivated. But the reasoning is probably something like:

1. If we start from Eq 34, for selecting `x_i`, `x_{i+1}`, it does not really matter what the ordering of `i` is (we could reorder the dimensions of all the parameters in the model to get either order).
2. Rotating half the array is probably more efficient than interleaved access (either when copying or when iterating with a stride).

Reply:
Found the original change through more spelunking:

https://github.com/EleutherAI/gpt-neox/pull/241

At any rate, I found out by first implementing the paper and then figuring out that it doesn't work the same way 🙈 .